flink-ml-lib/src/main/java/org/apache/flink/ml/classification/linear/LogisticRegression.java

```diff
@@ -0,0 +1,653 @@
+                                "genInitModel",
+                                PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO,
+                                new GenInitModel())
+                        .returns(PrimitiveArrayTypeInfo.DOUBLE_PRIMITIVE_ARRAY_TYPE_INFO);
+
+        DataStream<Tuple2<double[], double[]>> modelAndLoss = train(trainData, initModel);
+
+        DataStream<LogisticRegressionModelData> modelData =
+                modelAndLoss
+                        .connect(distinctLabelValues)
+                        .transform(
+                                "composeModelData",
+                                TypeInformation.of(LogisticRegressionModelData.class),
+                                new ComposeModelDataOp())
+                        .setParallelism(1);
+
+        LogisticRegressionModel model =
+                new LogisticRegressionModel().setModelData(tEnv.fromDataStream(modelData));
+        ReadWriteUtils.updateExistingParams(model, paramMap);
+        return model;
+    }
+
+    /** Pre-processes the training data. */
+    private static class PreprocessDataOp
+            extends AbstractUdfStreamOperator<
+                    Tuple3<Double, Double, double[]>,
+                    RichMapFunction<
+                            Tuple3<Double, Double, double[]>, Tuple3<Double, Double, double[]>>>
+            implements OneInputStreamOperator<
+                    Tuple3<Double, Double, double[]>, Tuple3<Double, Double, double[]>> {
+        public PreprocessDataOp(
+                RichMapFunction<Tuple3<Double, Double, double[]>, Tuple3<Double, Double, double[]>>
+                        userFunction) {
+            super(userFunction);
+        }
+
+        @Override
+        public void processElement(StreamRecord<Tuple3<Double, Double, double[]>> streamRecord)
+                throws Exception {
+            streamRecord.replace(userFunction.map(streamRecord.getValue()));
+            output.collect(streamRecord);
+        }
+    }
+
+    /** Pre-processes one training sample. */
+    private static class PreprocessOneRecord
+            extends RichMapFunction<
+                    Tuple3<Double, Double, double[]>, Tuple3<Double, Double, double[]>> {
+
+        String broadcastLabelsName;
+
+        double[] distinctLabelValues;
+
+        public PreprocessOneRecord(String broadcastLabelsName) {
+            this.broadcastLabelsName = broadcastLabelsName;
+        }
+
+        @Override
+        public Tuple3<Double, Double, double[]> map(Tuple3<Double, Double, double[]> value) {
+            if (distinctLabelValues == null) {
+                List<Double> labelList =
+                        getRuntimeContext().getBroadcastVariable(broadcastLabelsName);
+                distinctLabelValues = labelList.stream().mapToDouble(Double::doubleValue).toArray();
+            }
+            // label mapping
+            value.f1 = Math.abs(value.f1 - distinctLabelValues[0]) < 1e-7 ? 1. : -1.;
```

Comment:
Hmm.. instead of hardcoding the label value as either `1` or `-1`, should we choose from the value provided by user?

Does Spark's logistic regression takes more than 2 label values?

And if we choose to only support one label value, should this value type be boolean?

And how is `1e-7` choose? It looks like a magic number which is inconsistent with the `TOLERANCE` used in `LogisticRegressionTest`. Should we make it a `private static final` variable and share it across algorithms?

Reply:
Yes, we can directly use the user input labels, by restricting the input labels must be zero or one for binary classification (As Spark does).

Spark supports binary classification and multinomial classification. But currently, we only support binary classification in FlinkML.

My suggestion is that we directly use the labels from user input, meanwhile we make it `double`, such that we can add `multinomial classification` in the future.

What do you think?

Reply:
Would it be useful to specify add the `family` parameter following Spark's approach? We can choose to support only `binomial` for now.

Since logistic regression is used for classification, and there is always an integer number of classes, should the label type be `int` instead of `double`?

Reply:
Sure, we can add a parameter to let users specify `bionomial` or `multinomial` or `auto`.
For the name of parameter, I would suggest go with `HasMultiClass`, the default value could be `auto, bionomial, multinomial` following Sklearn [1] and I think it is more intuitive than `family`.

For the label type, personally I think using `int` is enough. But I find that both Spark and Sklearn use `double` as the label [2][3]. I guess it is because for usability --- Users may specify labels as `0.0`, `1.0`, though they are integers.

[1] https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html
[2] https://github.com/scikit-learn/scikit-learn/blob/0d378913be6d7e485b792ea36e9268be31ed52d0/sklearn/linear_model/_stochastic_gradient.py#L1402
[3] https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/regression/LabeledPoint.scala#L37

Reply:
Thanks for the detailed explanation! Sounds good. Let's use `HasMultiClass` and `double` value as suggested above.

Reply:
Hi both,
I'm trying flink-ml but I don't know, when will it support multiple labels?.
Thanks for info.

Reply:
Hi @Mrmachiner, I believe we will start to support multiple labels for LR as well as other training algorithms in the 2nd half of 2023, but we don't have concrete time or plan yet.

Our focus in the past few months has been on developing feature engineering algorithms in Flink ML, as that seems to be an easier way to help Flink ML get users in production.